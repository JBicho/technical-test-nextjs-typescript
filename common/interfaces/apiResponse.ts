interface PaginationQuery {
  page: string;
  limit: string;
  name?: string;
  powerThreshold?: string;
}

interface PaginatedResponse<T> {
  data: T[];
  countData: {
    count: number;
    min: number;
    max: number;
  };
  pagination: {
    currentPage: number;
    totalPages: number;
    totalItems: number;
    itemsPerPage: number;
    hasNextPage: boolean;
    hasPreviousPage: boolean;
  };
}

export type { PaginatedResponse, PaginationQuery };
