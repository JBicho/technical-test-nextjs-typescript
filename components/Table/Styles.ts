import styled from 'styled-components';

const StyledTable = styled.table`
  width: 100%;
  max-width: var(--global-max-width);
  min-width: var(--global-max-width);
  margin: 20px auto;
  border-spacing: 0;
  border-radius: 12px;
  background: rgba(255, 255, 255, 0.08);
  box-shadow: 0 4px 10px rgba(0, 0, 0, 0.1);
  backdrop-filter: blur(10px);
  overflow: hidden;
`;

const TableHeader = styled.thead`
  background: rgba(255, 255, 255, 0.15);

  th {
    padding: calc(var(--global-rhythm) / 1.5);
    font-weight: 600;
    font-size: 1rem;
    text-align: left;
  }

  tr:first-child th:first-child {
    border-top-left-radius: 12px;
  }
  tr:first-child th:last-child {
    border-top-right-radius: 12px;
  }
`;

const TableBody = styled.tbody`
  tr {
    border-bottom: 1px rgba(255, 255, 255, 1);

    &:hover {
      background: rgba(255, 255, 255, 0.12);
      cursor: pointer;
    }
  }

  td {
    padding: calc(var(--global-rhythm) / 1.5);
    font-size: 0.95rem;
    text-align: left;
    background: rgba(255, 255, 255, 0.05);
  }
`;

const EmptyListTd = styled.td`
  text-align: center !important;
`;

const TableFooter = styled.tfoot`
  td {
    padding: 12px;
    font-weight: 500;
    text-align: center;
    background: rgba(255, 255, 255, 0.1);

    span {
      font-size: 0.9rem;
    }

    button {
      background-color: transparent;
      height: 46px;
      width: 50px;

      &:first-child {
        margin-right: var(--global-rhythm);
      }

      &:last-child {
        margin-left: var(--global-rhythm);
      }
    }

    &:first-child {
      border-bottom-left-radius: 12px;
    }
    &:last-child {
      border-bottom-right-radius: 12px;
    }
  }
`;

export { StyledTable, TableBody, TableFooter, TableHeader, EmptyListTd };
