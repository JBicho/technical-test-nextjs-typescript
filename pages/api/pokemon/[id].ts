import { NextApiRequest, NextApiResponse } from 'next';
import fetchPokemonById from '../handlers/fetchByIdHandler';
import { errorMiddleware } from '../middleware/errorMiddleware';

const getPokemonById = async (req: NextApiRequest, res: NextApiResponse) => {
  return errorMiddleware(['GET'], fetchPokemonById)(req, res);
};

export default getPokemonById;
