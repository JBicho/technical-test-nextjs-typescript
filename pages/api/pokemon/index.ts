import { NextApiRequest, NextApiResponse } from 'next';
import { dataFecthHandler } from '../handlers/fetchDataHandler';
import { errorMiddleware } from '../middleware/errorMiddleware';

const getPokemonData = async (req: NextApiRequest, res: NextApiResponse) =>
  errorMiddleware(['GET'], dataFecthHandler)(req, res);

export default getPokemonData;
