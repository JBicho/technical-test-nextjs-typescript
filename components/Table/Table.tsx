import { useRouter } from 'next/router';
import { useState } from 'react';
import { Pagination } from '../../common/interfaces/pagination';
import { Pokemon } from '../../common/interfaces/pokemon';
import { calculatePokemonPower } from '../../common/utils/calculatePokemonPower';
import {
  EmptyListTd,
  StyledTable,
  TableBody,
  TableFooter,
  TableHeader,
} from './Styles';

export interface TableProps {
  pokemonList: Pokemon[];
  paginationData: Pagination;
  handlePagination: (page: number) => void;
}

export const Table = ({
  pokemonList = [],
  handlePagination,
  paginationData: {
    currentPage,
    hasNextPage,
    hasPreviousPage,
    itemsPerPage,
    totalItems,
    totalPages,
  },
}: TableProps) => {
  const router = useRouter();
  const [isNavigating, setIsNavigating] = useState(false);

  const handleRowClick = async (id: number) => {
    if (isNavigating) return;

    try {
      setIsNavigating(true);
      await router.push(`/pokemon/${id}`);
    } catch (err: unknown) {
      const error = err as Error;

      if (error?.message !== 'Loading initial props cancelled') {
        console.error('Navigation error:', error);
      }
    } finally {
      setIsNavigating(false);
    }
  };

  const handlePageChange = async (page: number) => {
    if (isNavigating) return;

    try {
      setIsNavigating(true);

      await handlePagination(page);
    } catch (err: unknown) {
      const error = err as Error;

      if (error?.message !== 'Loading initial props cancelled') {
        console.error('Pagination error:', error);
      }
    } finally {
      setIsNavigating(false);
    }
  };

  if (router.isFallback) {
    return <div>Loading...</div>;
  }

  return (
    <StyledTable>
      <TableHeader>
        <tr>
          <th>ID</th>
          <th>Name</th>
          <th>Type</th>
          <th colSpan={6}>Stats</th>
          <th>Power</th>
        </tr>
      </TableHeader>
      <TableBody>
        {!pokemonList.length && (
          <tr>
            <EmptyListTd colSpan={10}>No pokemon found</EmptyListTd>
          </tr>
        )}
        {pokemonList.map(
          (
            {
              id,
              name,
              type,
              hp,
              speed,
              attack,
              special_attack,
              defense,
              special_defense,
            },
            index
          ) => (
            <tr
              key={`${name}${index}`}
              onClick={() => handleRowClick(id)}
              style={{ cursor: isNavigating ? 'wait' : 'pointer' }}
            >
              <td>{id}</td>
              <td>{name}</td>
              <td>{type.join(', ')}</td>
              <td>{hp}</td>
              <td>{speed}</td>
              <td>{attack}</td>
              <td>{special_attack}</td>
              <td>{defense}</td>
              <td>{special_defense}</td>
              <td>
                {calculatePokemonPower({
                  attack,
                  defense,
                  hp,
                  special_attack,
                  special_defense,
                  speed,
                })}
              </td>
            </tr>
          )
        )}
      </TableBody>
      <TableFooter>
        <tr>
          <td colSpan={10}>
            <button
              onClick={() => handlePageChange(currentPage - 1)}
              disabled={!hasPreviousPage || currentPage === 1 || isNavigating}
            >
              <strong>Prev</strong>
            </button>
            <span>
              Page {currentPage} of {totalPages}
            </span>
            <button
              onClick={() => handlePageChange(currentPage + 1)}
              disabled={
                !hasNextPage || currentPage === totalPages || isNavigating
              }
            >
              <strong>Next</strong>
            </button>
          </td>
        </tr>
      </TableFooter>
    </StyledTable>
  );
};
