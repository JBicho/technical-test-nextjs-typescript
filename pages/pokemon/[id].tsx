import Head from 'next/head';
import Image from 'next/image';
import { useRouter } from 'next/router';
import {
  MAX_ID,
  MAX_METER_VALUE,
  NAVIGATE_NEXT,
  NAVIGATE_PREV,
} from '../../common/constants';
import { Pokemon } from '../../common/interfaces/pokemon';
import { logger } from '../../common/utils/logger';
import { Card } from '../../components/Card/Card';
import { Layout } from '../../components/LayoutComponent/Layout';
import { Meter } from '../../components/Meter/Meter';
import {
  DetailView,
  DetailViewNavigation,
  Details,
  ImageContainer,
  NameAndImage,
  NavAndDetails,
  TypesList,
} from '../../components/Pokemon/Styles';
import { useEffect } from 'react';
import Link from 'next/link';

const shimmer = (w: number, h: number): string => `
<svg width="${w}" height="${h}" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
  <defs>
    <linearGradient id="g">
      <stop stop-color="#eee" offset="20%" />
      <stop stop-color="#f5f5f5" offset="50%" />
      <stop stop-color="#eee" offset="70%" />
    </linearGradient>
  </defs>
  <rect width="${w}" height="${h}" fill="#eee" />
  <rect id="r" width="${w}" height="${h}" fill="url(#g)" />
</svg>`;

const toBase64 = (str: string): string =>
  typeof window === 'undefined'
    ? Buffer.from(str).toString('base64')
    : window.btoa(str);
const getBaseUrl = () => {
  if (typeof window !== 'undefined') {
    return '';
  }
  return process.env.BASE_URL || 'http://localhost:3000';
};

interface PokemonPageProps extends Pokemon {
  error?: string;
}

const PokemonPage = (props: PokemonPageProps) => {
  const router = useRouter();
  const { id } = router.query;

  if (props.error) {
    return (
      <Card>
        <DetailView>
          <h1>Error</h1>
          <p>{props.error}</p>
        </DetailView>
      </Card>
    );
  }

  const handleNavigation = async (direction: 'next' | 'prev') => {
    try {
      let newId = Number(id as string);

      if (direction === 'next') {
        newId += 1;
      }

      if (direction === 'prev') {
        newId -= 1;
      }

      if (newId < 1 || newId > MAX_ID) {
        return;
      }

      await router.push(`/pokemon/${newId}`);
    } catch (error) {
      logger.error('Navigation error:', error);
    }
  };

  return (
    <>
      <Head>
        <title>{props.name} - Pokemon Details</title>
        <meta
          name="description"
          content={`Details for ${props.name}, a ${props.type.join('/')} type Pokemon`}
        />
      </Head>
      <Card>
        <DetailView>
          <NameAndImage>
            <h1>{props.name}</h1>

            <ImageContainer>
              <Image
                src={`/images/${props.name.toLowerCase()}.jpg`}
                alt={`${props.name} Pokemon`}
                width={250}
                height={250}
                quality={85}
                priority
                placeholder="blur"
                blurDataURL={`data:image/svg+xml;base64,${toBase64(shimmer(250, 250))}`}
              />
            </ImageContainer>
            <h2>Type</h2>
            <TypesList>
              {props.type.map((type) => (
                <li key={type}>
                  <strong>{type}</strong>
                </li>
              ))}
            </TypesList>
          </NameAndImage>

          <NavAndDetails>
            <Details>
              <Meter
                id="hp-meter"
                label="HP"
                value={props.hp}
                max={MAX_METER_VALUE}
              />
              <Meter
                id="attack-meter"
                label="Attack"
                value={props.attack}
                max={MAX_METER_VALUE}
              />
              <Meter
                id="defense-meter"
                label="Defense"
                value={props.defense}
                max={MAX_METER_VALUE}
              />
              <Meter
                id="special_attack-meter"
                label="Special Attack"
                value={props.special_attack}
                max={MAX_METER_VALUE}
              />
              <Meter
                id="special_defense-meter"
                label="Special Defense"
                value={props.special_defense}
                max={MAX_METER_VALUE}
              />
              <Meter
                id="speed-meter"
                label="Speed"
                value={props.speed}
                max={MAX_METER_VALUE}
              />
            </Details>

            <DetailViewNavigation>
              <Link
                href={`/pokemon/${Number(id) - 1}`}
                prefetch={true}
                aria-label="View previous Pokemon"
              >
                <button
                  disabled={Number(id) === 1}
                  aria-label="View previous Pokemon"
                >
                  Previous
                </button>
              </Link>
              <Link
                href={`/pokemon/${Number(id) + 1}`}
                prefetch={true}
                aria-label="View next Pokemon"
              >
                <button
                  disabled={Number(id) === MAX_ID}
                  aria-label="View next Pokemon"
                >
                  Next
                </button>
              </Link>
            </DetailViewNavigation>
          </NavAndDetails>
        </DetailView>
      </Card>
    </>
  );
};

PokemonPage.getLayout = Layout;

export async function getServerSideProps({
  params: { id },
}: {
  params: { id: string };
}) {
  try {
    if (!id || isNaN(Number(id)) || Number(id) < 1 || Number(id) > MAX_ID) {
      return {
        props: {
          error: 'Invalid Pokemon ID',
        },
      };
    }

    const baseUrl = getBaseUrl();
    const res = await fetch(`${baseUrl}/api/pokemon/${id}`);

    if (!res.ok) {
      logger.error('Failed to fetch Pokemon data:', {
        statusCode: res.status,
        statusText: res.statusText,
        id,
      });

      return {
        props: {
          error: 'Pokemon not found',
        },
      };
    }

    const pokemon: Pokemon = await res.json();

    if (!pokemon) {
      return {
        props: {
          error: 'Pokemon data is invalid',
        },
      };
    }

    return {
      props: {
        ...pokemon,
      },
    };
  } catch (error) {
    logger.error('Error while fetching Pokemon data:', {
      error,
      id,
      url: process.env.BASE_URL,
    });

    return {
      props: {
        error: 'Failed to load Pokemon data',
      },
    };
  }
}

export default PokemonPage;
