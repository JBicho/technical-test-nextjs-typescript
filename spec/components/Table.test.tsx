import {
  act,
  fireEvent,
  render,
  screen,
  waitFor,
} from '@testing-library/react';
import { useRouter } from 'next/router';
import { Pokemon } from '../../common/interfaces/pokemon';
import { Table } from '../../components/Table/Table';
import '../setupTests';

jest.mock('next/router', () => ({
  useRouter: jest.fn(),
}));

describe('Test Table Component', () => {
  beforeEach(() => {
    (useRouter as jest.Mock).mockReturnValue({
      isFallback: false,
      push: jest.fn(),
    });
  });

  it('Should match snapshot with empty list', () => {
    const paginationData = {
      currentPage: 1,
      hasNextPage: false,
      hasPreviousPage: false,
      itemsPerPage: 10,
      totalItems: 0,
      totalPages: 1,
    };

    const { asFragment } = render(
      <Table
        pokemonList={[]}
        paginationData={paginationData}
        handlePagination={() => {}}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('Should match snapshot with non empty list', () => {
    const handlePagination = jest.fn();

    const paginationData = {
      currentPage: 1,
      hasNextPage: true,
      hasPreviousPage: false,
      itemsPerPage: 10,
      totalItems: 20,
      totalPages: 2,
    };

    const pokemonList = [
      {
        id: 1,
        name: 'Pikachu',
        type: ['Electric'],
        hp: 35,
        attack: 55,
        defense: 40,
        special_attack: 50,
        special_defense: 50,
        speed: 90,
      },
    ];

    const { asFragment } = render(
      <Table
        pokemonList={pokemonList}
        paginationData={paginationData}
        handlePagination={handlePagination}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('Should handle row click and navigate', async () => {
    const mockPush = jest.fn();

    (useRouter as jest.Mock).mockReturnValue({
      push: mockPush,
      isFallback: false,
    });

    const pokemonList = [
      {
        id: 1,
        name: 'Pikachu',
        type: ['Electric'],
        hp: 35,
        speed: 90,
        attack: 55,
        special_attack: 50,
        defense: 40,
        special_defense: 50,
      },
    ];

    const paginationData = {
      currentPage: 1,
      hasNextPage: true,
      hasPreviousPage: false,
      itemsPerPage: 10,
      totalItems: 1,
      totalPages: 1,
    };

    const handlePagination = jest.fn();

    render(
      <Table
        pokemonList={pokemonList}
        paginationData={paginationData}
        handlePagination={handlePagination}
      />
    );

    const row = screen.getByText('Pikachu').closest('tr');

    if (row) {
      await act(async () => {
        fireEvent.click(row);
      });
    } else {
      throw new Error('Row not found');
    }

    await waitFor(() => {
      expect(mockPush).toHaveBeenCalledWith('/pokemon/1');
    });
  });

  it('Should handle page change and trigger pagination', async () => {
    const handlePagination = jest.fn();

    const paginationData = {
      currentPage: 1,
      hasNextPage: true,
      hasPreviousPage: false,
      itemsPerPage: 10,
      totalItems: 20,
      totalPages: 2,
    };

    const pokemonList = [
      {
        id: 1,
        name: 'Pikachu',
        type: ['Electric'],
        hp: 35,
        attack: 55,
        defense: 40,
        special_attack: 50,
        special_defense: 50,
        speed: 90,
      },
    ];

    render(
      <Table
        pokemonList={pokemonList}
        paginationData={paginationData}
        handlePagination={handlePagination}
      />
    );

    const nextPageButton = screen.getByText('Next');

    fireEvent.click(nextPageButton);

    await waitFor(() => {
      expect(handlePagination).toHaveBeenCalledWith(2);
    });
  });

  it('Should display "No pokemon found" when the pokemon list is empty', () => {
    const pokemonList: Pokemon[] = [];
    const paginationData = {
      currentPage: 1,
      hasNextPage: true,
      hasPreviousPage: false,
      itemsPerPage: 10,
      totalItems: 0,
      totalPages: 1,
    };
    const handlePagination = jest.fn();

    render(
      <Table
        pokemonList={pokemonList}
        paginationData={paginationData}
        handlePagination={handlePagination}
      />
    );

    expect(screen.getByText('No pokemon found')).toBeInTheDocument();
  });

  it('Should display loading state when router is in fallback mode', () => {
    (useRouter as jest.Mock).mockReturnValue({
      isFallback: true,
      push: jest.fn(),
    });

    const pokemonList: Pokemon[] = [];
    const paginationData = {
      currentPage: 1,
      hasNextPage: false,
      hasPreviousPage: false,
      itemsPerPage: 10,
      totalItems: 0,
      totalPages: 1,
    };

    const handlePagination = jest.fn();

    render(
      <Table
        pokemonList={pokemonList}
        paginationData={paginationData}
        handlePagination={handlePagination}
      />
    );

    expect(screen.getByText('Loading...')).toBeInTheDocument();
  });
});
