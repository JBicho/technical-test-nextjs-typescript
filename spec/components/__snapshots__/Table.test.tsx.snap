// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Table Component Should match snapshot with empty list 1`] = `
<DocumentFragment>
  <table
    class="Styles__StyledTable-sc-1o1cw3u-0 bKkKXR"
  >
    <thead
      class="Styles__TableHeader-sc-1o1cw3u-1 jSvCgs"
    >
      <tr>
        <th>
          ID
        </th>
        <th>
          Name
        </th>
        <th>
          Type
        </th>
        <th
          colspan="6"
        >
          Stats
        </th>
        <th>
          Power
        </th>
      </tr>
    </thead>
    <tbody
      class="Styles__TableBody-sc-1o1cw3u-2 fykNoR"
    >
      <tr>
        <td
          class="Styles__EmptyListTd-sc-1o1cw3u-3 bgUgKv"
          colspan="10"
        >
          No pokemon found
        </td>
      </tr>
    </tbody>
    <tfoot
      class="Styles__TableFooter-sc-1o1cw3u-4 iiILBb"
    >
      <tr>
        <td
          colspan="10"
        >
          <button
            disabled=""
          >
            <strong>
              Prev
            </strong>
          </button>
          <span>
            Page 1 of 1
          </span>
          <button
            disabled=""
          >
            <strong>
              Next
            </strong>
          </button>
        </td>
      </tr>
    </tfoot>
  </table>
</DocumentFragment>
`;

exports[`Test Table Component Should match snapshot with non empty list 1`] = `
<DocumentFragment>
  <table
    class="Styles__StyledTable-sc-1o1cw3u-0 bKkKXR"
  >
    <thead
      class="Styles__TableHeader-sc-1o1cw3u-1 jSvCgs"
    >
      <tr>
        <th>
          ID
        </th>
        <th>
          Name
        </th>
        <th>
          Type
        </th>
        <th
          colspan="6"
        >
          Stats
        </th>
        <th>
          Power
        </th>
      </tr>
    </thead>
    <tbody
      class="Styles__TableBody-sc-1o1cw3u-2 fykNoR"
    >
      <tr
        style="cursor: pointer;"
      >
        <td>
          1
        </td>
        <td>
          Pikachu
        </td>
        <td>
          Electric
        </td>
        <td>
          35
        </td>
        <td>
          90
        </td>
        <td>
          55
        </td>
        <td>
          50
        </td>
        <td>
          40
        </td>
        <td>
          50
        </td>
        <td>
          320
        </td>
      </tr>
    </tbody>
    <tfoot
      class="Styles__TableFooter-sc-1o1cw3u-4 iiILBb"
    >
      <tr>
        <td
          colspan="10"
        >
          <button
            disabled=""
          >
            <strong>
              Prev
            </strong>
          </button>
          <span>
            Page 1 of 2
          </span>
          <button>
            <strong>
              Next
            </strong>
          </button>
        </td>
      </tr>
    </tfoot>
  </table>
</DocumentFragment>
`;
